### PR TITLE
Babbage: Reference scripts

### LINKED ISSUE #2465: Babbage TxOut
Implement the new Babbage `TxOut` type and all necessary instances.

Babbage adds two things to the Alonzo `TxOut' type.

- [x] [Inline datums](https://github.com/cardano-foundation/CIPs/pull/160)
- [x] [reference scripts](https://github.com/cardano-foundation/CIPs/pull/161)

### PR DESCRIPTION
**Pull request by goolord**
resolves https://github.com/input-output-hk/cardano-ledger/issues/2465

### CODE REVIEW

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/TxBody.hs [hunk 1]
```diff
@@ -163,6 +164,11 @@ data TxOut era
       {-# UNPACK #-} !(CompactAddr (Crypto era))
       !(CompactForm (Core.Value era))
       {-# UNPACK #-} !(BinaryData era) -- Inline data
+  | TxOutCompactRefScript'
+      {-# UNPACK #-} !(CompactAddr (Crypto era))
+      !(CompactForm (Core.Value era))
+      !(Datum era)
```
Comment (goolord) [inline eras/babbage/impl/src/Cardano/Ledger/Babbage/TxBody.hs:170]: this is kind of a compromise between performance and having a nice API + not polluting the CBOR with a bunch of extra constructors to represent every possible new permutation of `Addr × Value × (Datum | DataHash)? × Script?`

**Comment by goolord** [comment]
https://github.com/input-output-hk/cardano-ledger/runs/5059981296?check_suite_focus=true#step:20:2261 this failure seems unrelated

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/TxBody.hs (hunk 1 above)
Comment (JaredCorduan) [reply eras/babbage/impl/src/Cardano/Ledger/Babbage/TxBody.hs:170]: yep, I like this. we are indeed on a combinatorially explosive path that has to be dampened somehow.

**Review by JaredCorduan — APPROVED** [review]
Looks good to me! We are still trying to decide on the final serialization schema for the Babbage era transaction outputs, but I think this PR does not need to be blocked by that decision.

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/TxBody.hs (hunk 1 above)
Comment (lehins) [reply eras/babbage/impl/src/Cardano/Ledger/Babbage/TxBody.hs:170]: It is not compromise between performance. It is compromise between memory overhead. `Datum era` takes up more than twice what the `BinaryData era` does!

I am afraid that in that we don't get UTxO HD in time we will need to save every byte possible. In my opinion, once we don't have to worry about memory overhead that much we can figure out how we can dampen this, but until then we should be careful!

Comment (lehins) [reply eras/babbage/impl/src/Cardano/Ledger/Babbage/TxBody.hs:170]: My previous comment was wrong. This looks good, I like it too! :wink:

**Comment by JaredCorduan** [comment]
I'm not sure what is going on with ormolu, I'll try to restart the github action

File: eras/babbage/impl/src/Cardano/Ledger/Babbage/TxBody.hs (hunk 1 above)
Comment (yihuang) [reply eras/babbage/impl/src/Cardano/Ledger/Babbage/TxBody.hs:170]: Does this mean that inline data and ref script can't be used in the same output?

Comment (lehins) [reply eras/babbage/impl/src/Cardano/Ledger/Babbage/TxBody.hs:170]: On the contrary, it means that inline datum is optional with a ref script

Comment (goolord) [reply eras/babbage/impl/src/Cardano/Ledger/Babbage/TxBody.hs:170]: `TxOutCompactRefScript' caddr value (Datum inlineDatum) refScript` is valid

**Pull request merged**